linsolve/linsolve.py

```diff
@@ -323,78 +340,125 @@ def get_weighted_data(self):
+        At = A.T.conj()
+        AtA = At.dot(A).toarray()
+        try: AtAi = np.linalg.pinv(AtA, rcond=rcond)
+        except(np.linalg.LinAlgError): AtAi = np.linalg.inv(AtA)
+        return AtAi.dot(At.dot(y)) 
+
+    def _invert_solve_sparse(self, A, y, rcond):
+        '''Use linalg.solve to solve a fully constrained (non-degenerate) system of equations.
+        Tends to be ~3x slower than lsqr for sparse matrices.'''
+        At = A.T.conj()
+        AtA = At.dot(A).toarray()
+        Aty = At.dot(y) # automatically dense bc y is dense
+        #xhat = scipy.sparse.linalg.spsolve(AtA, Aty)
+        return np.linalg.solve(AtA, Aty)
+
+    def _invert_lsqr(self, A, y, rcond):
+        '''Use np.linalg.lstsq to solve a system of equations.  Usually the best 
+        performer, but for a fully-constrained system, 'solve' can be faster.  Also,
+        there are a couple corner cases where lstsq is unstable but pinv works 
+        for the same rcond. It seems particularly the case for single precision matrices.'''
+        xhat = np.linalg.lstsq(A, y[...,0], rcond=rcond)[0]
+        xhat.shape += (1,)
+        return xhat
+
+    def _gen_AtAiAt(self, A, rcond):
+        '''Helper function for forming (At A)^-1 At.  Uses pinv to invert.'''
+        At = A.T.conj()
+        AtA = At.dot(A)
+        # finding inverse is about 3x slower than solve
+        try: AtAi = np.linalg.pinv(AtA, rcond=rcond)
+        except(np.linalg.LinAlgError): 
+            AtAi = np.linalg.inv(AtA)
+        AtAiAt = AtAi.dot(A.T.conj()) 
+        return AtAiAt
+
+    def _invert_pinv(self, A, y, rcond):
+        '''Use np.linalg.pinv to invert AtA matrix.  Tends to be about ~3x slower than solve.'''
+        AtAiAt = self._gen_AtAiAt(A, rcond)
+        return np.dot(AtAiAt,y)
+
+    def _invert_solve(self, A, y, rcond):
+        '''Use np.linalg.solve to solve a system of equations.  Requires a fully constrained
+        system of equations (i.e. doesn't deal with singular matrices).  Can by ~1.5x faster that lstsq
+        for this case.'''
+        At = A.T.conj()
+        AtA = At.dot(A)
+        Aty = At.dot(y)
+        return np.linalg.solve(AtA, Aty) # is supposed to error if singular, but doesn't seem to.
+
+    def _invert_default(self, A, y, rcond):
+        '''Use the lsqr inverter, but if an LinAlgError is encountered, try using pinv.'''
+        try:
+            xhat = self._invert_lsqr(A, y, rcond)
+        except(np.linalg.LinAlgError):
+            xhat = self._invert_pinv(A, y, rcond)
+        return xhat
+
+    def solve(self, rcond=1e-10, mode='default', verbose=False):
         """Compute x' = (At A)^-1 At * y, returning x' as dict of prms:values.
 
         Args:
             rcond: cutoff ratio for singular values useed in numpy.linalg.lstsq, numpy.linalg.pinv,
                 or (if sparse) as atol and btol in scipy.sparse.linalg.lsqr
+            mode: 'default', 'lsqr', 'pinv', or 'solve', selects what inverter to use.  'default'
```

Comment:
I'd expand on these here and explain in more detail what they do (rather than just when to use them).

Reply:
Done.